## Ques3. Analyze data4 using linear regression

### (a) Describe the cross validation techniques used.


In [188]:
df4=pd.read_csv("data4.csv",header=None)
df4

0      1      2      3      4      5      6      7      8      9    \
0    47836  20431  74759  10681  58511  44219  38087  14319  26694  36772   
1    63150   1643   8632   1725  28803  21985  72649  50667  18369  76683   
2    29413  73895  14579  44788  36337  19321  44175  49921  74890  63204   
3    16483  52296   7928  24066  30912  19452   2636  26236  54656  36150   
4     6934  74610  39182  75153  62045  12333   4310  64238  76970  26675   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
145  26067  49898  38046  12520  15018  10705  57251  79947  64433  37393   
146  43716  18916  28997  44965  25755  53718  46220  28433  56069  52536   
147  31911  14170  63050  55585  32309  45680  34664   3767  69779  23215   
148  33208  66372  62424  34117  43886  13582  70740  17093   4176  60363   
149  14460  61354  53481  66902   3900  11813  31445  31828  17575  44650   

     ...    191    192    193    194    195    196    197    198    199  \
0    ...  31070     42  22402  68958  63416  33276  54156  47900  75036   
1    ...  23519  54290  72179  23853  30416  74502  68129  45171   2777   
2    ...  71859  61987  35199  24264   6344  11770  61554   9932  30253   
3    ...  30880  72951  19416  17689  34736   1588  46116  45362  14261   
4    ...    967  28340   3480  11187  62838  33922  21504  37827   2788   
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
145  ...  77787  13526  61179   2675  67725  27647  69518  19906  40329   
146  ...  70457   3314  71703  25889  40190  65019  74417   2760  18703   
147  ...  56367  78590  15092  51605  18322    397  26997  10982  78800   
148  ...   6669  46792  53882  46743  56964  19019  21694  20013  48451   
149  ...  35343  49451  41272  53775  34575   2550  42090  30419  38293   

           200  
0    -118500.0  
1    -780230.0  
2    -903360.0  
3    -449080.0  
4    -398170.0  
..         ...  
145 -1139400.0  
146  -624190.0  
147  -506500.0  
148  -290900.0  
149 -1409300.0  

[150 rows x 201 columns]

In [84]:
len(df4)

150

In [85]:
len(df4.columns)

201


####  Data Set 4 :
####  Cardinality or Data Set length : N = 150
####  Number of attributes :  n = 200
####  Dimensions :       x = N*(n+1) = 150 * 201        ;       w= (n+1) * 1 = 201 * 1       ;        y_pred = N * 1 = 150 * 1

####  K_Fold Cross Validation Technique is used with Number of Folds k = 5
####  From the given data set length = 150 
####  80% of data is used to Train the model
####  20% of data is used to Test the model

####  TRAIN THE MODEL :
#### Linear Regression is Performed by using Iterative techniques called Batch Gradient Descent from the 80%  Data = Train Data=120
####  For Each Fold i,e 1 to k (i,e 5 )  The Parameters W and W0 are stored in a list  len(W) = len(W0) = 5
####  Now we found Parameters (W,W0) in each fold , which will be used in testing the 20% Data

#### TESTING THE MODEL :
#### WIth available 20% Data = Test Data = 30
#### Find y_predicted i,e y_pred = (w)Transpose * x + w0
#### As length of x =30 ,SO  lenght of Y_pred = 30
#### Now Compare this Y_pred with Y_test
####  So that we can see the performance of the model
####  How well the Model is performing For the Test Data With which the Model hasn't Trained

#### The Output Results of both Y_pred and Y_train are given in the Output of the Program to compare

In [189]:
df4_normalized=normalisation(df4)
df4_normalized

0         1         2         3         4         5         6    \
0    0.602099  0.248759  0.935526  0.123658  0.740695  0.558660  0.477498   
1    0.797408  0.010864  0.090604  0.010024  0.363943  0.277146  0.912103   
2    0.367139  0.925724  0.166590  0.556411  0.459488  0.243416  0.554052   
3    0.202234  0.652236  0.081608  0.293488  0.390689  0.245075  0.031713   
4    0.080450  0.934778  0.480949  0.941685  0.785512  0.154938  0.052763   
..        ...       ...       ...       ...       ...       ...       ...   
145  0.324465  0.621872  0.466434  0.146992  0.189124  0.134325  0.718478   
146  0.549554  0.229576  0.350813  0.558657  0.325289  0.678931  0.579767   
147  0.398998  0.169482  0.785917  0.693405  0.408406  0.577159  0.434455   
148  0.415539  0.830467  0.777918  0.421017  0.555223  0.170752  0.888098   
149  0.176434  0.766929  0.663651  0.836996  0.048128  0.148354  0.393977   

          7         8         9    ...       191       192       193  \
0    0.178792  0.331300  0.463069  ...  0.381181  0.000000  0.275047   
1    0.633534  0.226138  0.981879  ...  0.285566  0.680226  0.906407   
2    0.624201  0.940112  0.806663  ...  0.897674  0.776740  0.437361   
3    0.327883  0.684516  0.454984  ...  0.378775  0.914219  0.237174   
4    0.803318  0.966386  0.331817  ...  0.000000  0.354834  0.035045   
..        ...       ...       ...  ...       ...       ...       ...   
145  0.999850  0.808019  0.471142  ...  0.972738  0.169078  0.766885   
146  0.355369  0.702365  0.667988  ...  0.879921  0.041028  0.900369   
147  0.046778  0.875549  0.286840  ...  0.701506  0.984928  0.182329   
148  0.213497  0.046852  0.769733  ...  0.072202  0.586207  0.674332   
149  0.397843  0.216108  0.565477  ...  0.435288  0.619549  0.514390   

          194       195       196       197       198       199       200  
0    0.861336  0.793197  0.415464  0.673492  0.588960  0.951948  0.858113  
1    0.294811  0.380341  0.936402  0.850751  0.553877  0.034117  0.435865  
2    0.299974  0.079181  0.143711  0.767342  0.100864  0.383117  0.357296  
3    0.217391  0.434387  0.015050  0.571497  0.556333  0.179987  0.647171  
4    0.135725  0.785965  0.423627  0.259273  0.459467  0.034257  0.679656  
..        ...       ...       ...       ...       ...       ...       ...  
145  0.028813  0.847106  0.344335  0.868372  0.229084  0.511102  0.206680  
146  0.320384  0.502621  0.816574  0.930520  0.008665  0.236409  0.535434  
147  0.643380  0.229035  0.000000  0.328957  0.114362  0.999759  0.610531  
148  0.582313  0.712477  0.235310  0.261684  0.230460  0.614267  0.748105  
149  0.670636  0.432373  0.027206  0.520424  0.364234  0.485240  0.034457  

[150 rows x 201 columns]

In [190]:
KFold_cross_validation_train(df4_normalized,1000,0.0000001,5)

 Data Set Lenght : 150 
120
120
w0:0.0043845971511259985 w:[0.0018771  0.00235655 0.00224366 0.0024664  0.00214409 0.00222267
 0.00235117 0.00214979 0.00232338 0.00217056 0.0022228  0.00216988
 0.00225035 0.00235483 0.00229209 0.00236982 0.00233294 0.0023173
 0.00225091 0.00219959 0.00212299 0.00222118 0.00217256 0.00202903
 0.00209928 0.00227983 0.00236439 0.00188155 0.00242038 0.00222583
 0.00207504 0.00196696 0.00198086 0.0020227  0.00187339 0.00225444
 0.00230656 0.00177826 0.00249913 0.00202929 0.00218718 0.00224871
 0.00218533 0.00201726 0.00212362 0.00208177 0.00234851 0.00209726
 0.00232316 0.00196104 0.00209422 0.00225066 0.00213629 0.00220904
 0.0022563  0.00238172 0.00201813 0.00215366 0.00205609 0.00221831
 0.00218652 0.00208511 0.00221933 0.0021372  0.00252795 0.0023398
 0.00226029 0.00212506 0.00198163 0.00233458 0.00244717 0.0022078
 0.00190844 0.00174782 0.00217137 0.00226045 0.00215362 0.00215998
 0.00214363 0.00226996 0.00207575 0.00210224 0.00201103 0.00221988
 0.001

In [191]:
KFold_cross_validation_test(df4_normalized,10000,0.00000001,5)

 Data Set Lenght : 150 
 
Lenth of x_train:30
Lenth of y_train:30
 
For Kth Fold: 0 
 
 
 y_pred:1.1091949052525745 y_train:0.8581127644911815
 
 y_pred:1.0859013653961636 y_train:0.43586487659205186
 
 y_pred:1.095186920846579 y_train:0.35729600040838205
 
 y_pred:1.093404016923765 y_train:0.6471706781694275
 
 y_pred:1.0913867486874516 y_train:0.6796561933688966
 
 y_pred:1.1030940743612612 y_train:0.6035248474948314
 
 y_pred:1.104257769835828 y_train:0.7326948109956865
 
 y_pred:1.104032554918817 y_train:0.39836391944664234
 
 y_pred:1.0982429711168253 y_train:0.391714949335103
 
 y_pred:1.0794270442373117 y_train:0.4746356466474387
 
 y_pred:1.0976400271868065 y_train:0.5007274305112432
 
 y_pred:1.0902184654549933 y_train:0.4567561703974068
 
 y_pred:1.1026952140572823 y_train:0.6795094310727686
 
 y_pred:1.0988145330126247 y_train:0.1593328058398632
 
 y_pred:1.094324598578642 y_train:0.5622591184052681
 
 y_pred:1.101430002873572 y_train:0.3990658260802981
 
 y_pred:1.097746711

### (b) Describe the method used to find the parameters.


In [192]:
direct_method(df4_normalized)

array([ 0.02618643,  0.03644462,  0.07184135,  0.04279338,  0.11135849,
        0.0755525 , -0.0067849 , -0.00948461,  0.05941125, -0.05161995,
        0.08381129,  0.02011506,  0.03301432,  0.05208218,  0.10282182,
        0.02608357,  0.01496933,  0.13760615, -0.01331554, -0.04206195,
       -0.05203123,  0.03735075, -0.03592582,  0.05812446, -0.00060075,
       -0.00474081,  0.00106134, -0.0735455 ,  0.04367143,  0.02150794,
       -0.02832759, -0.0207988 , -0.02460646, -0.04998553, -0.08435516,
       -0.05741892,  0.01338688, -0.11140435,  0.01319061, -0.04717151,
       -0.00120886,  0.05749328,  0.01775728,  0.00749517, -0.00069784,
       -0.01956835,  0.05404813,  0.01670294,  0.10728236, -0.03301209,
       -0.05919926,  0.0085003 ,  0.04279652, -0.01132609, -0.03383341,
        0.10707248, -0.06814003, -0.04099271, -0.04348215, -0.00927299,
       -0.00360908, -0.01905231, -0.03249939,  0.02517207,  0.0187904 ,
       -0.01340945, -0.02539673,  0.05563285, -0.00401337,  0.11

### (c) Report the values of the parameters of the model.


In [199]:
gradient_descent(df4,10000,0.00001)

(0.028408488505592664,
 array([-0.01030574,  0.03418969,  0.02398565,  0.04415821,  0.0609246 ,
         0.04338255,  0.02005092, -0.01786572,  0.01701777, -0.02970881,
         0.0569712 , -0.02281897,  0.00769494,  0.03259814,  0.07382828,
         0.00513591,  0.00371343,  0.09597255,  0.00252242, -0.03071788,
        -0.0495393 ,  0.02784127, -0.00435424,  0.03088234, -0.02621302,
        -0.00768687,  0.03678709, -0.08301578,  0.02090787,  0.02933804,
        -0.02218743, -0.01581825, -0.05608965, -0.0354468 , -0.04614047,
        -0.00896815, -0.0076159 , -0.08186132,  0.0215497 , -0.01337704,
         0.00685541,  0.04511716,  0.0075846 , -0.00624826,  0.01309032,
        -0.03110099,  0.03796261,  0.02316843,  0.0535337 , -0.03234846,
        -0.02352478, -0.00129217,  0.03071414,  0.01344619,  0.00136381,
         0.06933062, -0.0532721 , -0.0228845 , -0.0144373 ,  0.00250756,
         0.00039794, -0.01782041, -0.00277282,  0.00719318,  0.04837018,
         0.01253204,  0.0093

### (d) Report the average mean square error and standard deviation.


In [200]:
avg_mse = np.mean(mse_list)
print(avg_mse)

0.01022965857027944


In [201]:
std=np.std(y-y_pred)
print(std)


0.04857330463414661
